packages/imperative/src/rest/src/client/AbstractRestClient.ts

```diff
@@ -459,63 +480,48 @@ export abstract class AbstractRestClient {
-                    AbstractSession.getBase64Auth(this.session.ISession.user,  this.session.ISession.password));
-                headerKeys.forEach((property) => {
-                    options.headers[property] = authentication;
-                });
-            } else if (this.session.ISession.tokenValue) {
-                this.log.trace("Using cookie authentication with token %s", this.session.ISession.tokenValue);
-                const headerKeys: string[] = Object.keys(Headers.COOKIE_AUTHORIZATION);
-                const authentication: string = `${this.session.ISession.tokenType}=${this.session.ISession.tokenValue}`;
-                headerKeys.forEach((property) => {
-                    options.headers[property] = authentication;
-                });
-            } else {
-                throw new ImperativeError({msg: "No credentials for a BASIC or TOKEN type of session."});
+        let credsAreSet: boolean = false;
+        for (const nextAuthType of this.session.ISession.authTypeOrder) {
+            if (nextAuthType === SessConstants.AUTH_TYPE_TOKEN) {
+                credsAreSet ||= this.setTokenAuth(options);
+
+            } else if (nextAuthType === SessConstants.AUTH_TYPE_BASIC) {
+                credsAreSet ||= this.setPasswordAuth(options);
+
+            } else if (nextAuthType === SessConstants.AUTH_TYPE_BEARER) {
+                credsAreSet ||= this.setBearerAuth(options);
+
+            } else if (nextAuthType === SessConstants.AUTH_TYPE_CERT_PEM) {
+                credsAreSet ||= this.setCertPemAuth(options);
             }
-        } else if (this.session.ISession.type === SessConstants.AUTH_TYPE_BEARER) {
-            this.log.trace("Using bearer authentication");
-            const headerKeys: string[] = Object.keys(Headers.BASIC_AUTHORIZATION);
-            const authentication: string = AbstractSession.BEARER_PREFIX + this.session.ISession.tokenValue;
-            headerKeys.forEach((property) => {
-                options.headers[property] = authentication;
-            });
-        } else if (this.session.ISession.type === SessConstants.AUTH_TYPE_CERT_PEM) {
-            this.log.trace("Using PEM Certificate authentication");
-            try {
-                // Doing this again for SDKs using certificates
-                options.cert = readFileSync(this.session.ISession.cert);
-                options.key = readFileSync(this.session.ISession.certKey);
-            } catch (err) {
-                throw new ImperativeError({
-                    msg: "Failed to open one or more PEM certificate files, the file(s) did not exist.",
-                    causeErrors: err,
-                    additionalDetails: err.message,
-                });
+
+            if (credsAreSet) {
+                break;
             }
+            // else if (this.session.ISession.type === SessConstants.AUTH_TYPE_CERT_PFX) {
```

Comment:
I know this is commented out code, but should we move this to its own little `setCertPfxAuth` function and have this `else if` commented out right below the `setCertPemAuth`?

Reply:
I thought about creating a function for setCertPfxAuth, but I am reluctant to do it for the following reasons:

1. Once the logic is moved into its own function, the conditionals must be changed to achieve the same result. I prefer not to make logic changes into commented code that I cannot test. I do not want to accidentally lose the conditions specified by the original programmer.

2. If a new function is created with a commented-out body, some code-scanning tool is going to tell us what horrible programmers we are.

3. If I instead comment out the whole function, a programmer would have to find 2 blocks of comments that are separated by many lines of code to determine how to update the program to handle the new credential type.

I think a reasonable compromise would be to add some instructional comments at the beginning of the existing commented-out code-block describing how a new function should be created to handle Pfx certs. When a programmer wants to add Pfx certs, all of the information will be in one place, located where the programmer will likely look. I plan to make this change.

Reply:
I guess I inclined to agree with the reasons. 😅 
It's definitely not idea to have to modify code that we won't be testing, plus tools will start complaining about it 😋 

Thanks for adding the information about the commented out code though 🥳